src/main/java/org/jenkinsci/plugins/workflow/support/pickles/serialization/RiverWriter.java

```diff
@@ -169,26 +164,72 @@ public ObjectOutput getObjectOutput() {
 
     public void close() throws IOException {
         marshaller.finish();
-        int ephemeralsOffset = dout.size();
+        int ephemeralsOffset = (int)channel.position();
 
         // write the ephemerals stream
         pickling = false;
-        marshaller.start(Marshalling.createByteOutput(dout));
+        marshaller.start(new FileChannelOutput(channel));
         marshaller.writeObject(pickles);
         marshaller.finish();
-        dout.close();
 
         // back fill the offset to the ephemerals stream
-        RandomAccessFile raf = new RandomAccessFile(file, "rw");
         try {
```

Comment:
This should probably be moved up, to ensure we always close the channel.

Reply:
Done in 9f5f6c4.